src/wasm/nxt_wasm.c

```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright (C) Andrew Clayton
+ * Copyright (C) F5, Inc.
+ */
+
+#include <nxt_main.h>
+#include <nxt_application.h>
+#include <nxt_unit.h>
+#include <nxt_unit_request.h>
+
+#include "nxt_wasm.h"
+
+#define NXT_WASM_VERSION    "0.1"
+
+static nxt_wasm_ctx_t               nxt_wasm_ctx;
+
+static const nxt_wasm_operations_t  *nxt_wops;
+
+#define NXT_WASM_DO_HOOK(hook)  nxt_wops->exec_hook(&nxt_wasm_ctx, hook);
+
+void
+nxt_wasm_do_response_end(nxt_wasm_ctx_t *ctx)
+{
+    nxt_unit_request_done(ctx->req, NXT_UNIT_OK);
+
+    NXT_WASM_DO_HOOK(NXT_WASM_FH_RESPONSE_END);
+}
+
+
+void
+nxt_wasm_do_send_headers(uint32_t offs, nxt_wasm_ctx_t *ctx)
+{
+    size_t                   fields_len;
+    unsigned int             i;
+    nxt_wasm_response_hdr_t  *rh;
+
+    rh = (nxt_wasm_response_hdr_t *)(ctx->baddr + offs);
+
+    fields_len = 0;
+    for (i = 0; i < rh->nr_fields; i++)
+        fields_len += rh->fields[i].name_len + rh->fields[i].value_len;
+
+    nxt_unit_response_init(ctx->req, 200, rh->nr_fields, fields_len);
+
+    for (i = 0; i < rh->nr_fields; i++) {
+        const char *name;
+        const char *val;
+
+        name = (const char *)(uint8_t *)rh + rh->fields[i].name_offs;
+        val = (const char *)(uint8_t *)rh + rh->fields[i].value_offs;
+
+        nxt_unit_response_add_field(ctx->req, name, rh->fields[i].name_len,
+                                    val, rh->fields[i].value_len);
+    }
+
+    nxt_unit_response_send(ctx->req);
+}
+
+
+void
+nxt_wasm_do_send_response(uint32_t offs, nxt_wasm_ctx_t *ctx)
+{
+    nxt_wasm_response_t      *resp;
+    nxt_unit_request_info_t  *req = ctx->req;
+
+    if (!nxt_unit_response_is_init(req)) {
+        nxt_unit_response_init(req, 200, 0, 0);
+    }
+
+    resp = (nxt_wasm_response_t *)(nxt_wasm_ctx.baddr + offs);
+
+    nxt_unit_response_write(req, (const char *)resp->data, resp->size);
+}
+
+
+static void
+nxt_wasm_request_handler(nxt_unit_request_info_t *req)
+{
+    size_t                     offset, read_bytes, content_sent, content_len;
+    ssize_t                    bytes_read;
+    nxt_unit_field_t           *sf, *sf_end;
+    nxt_unit_request_t         *r;
+    nxt_wasm_request_t         *wr;
+    nxt_wasm_http_hdr_field_t  *df;
+
+    NXT_WASM_DO_HOOK(NXT_WASM_FH_REQUEST_INIT);
+
+    wr = (nxt_wasm_request_t *)nxt_wasm_ctx.baddr;
+
+#define SET_REQ_MEMBER(dmember, smember) \
+    do { \
+        const char *str = nxt_unit_sptr_get(&r->smember); \
+        wr->dmember##_offs = offset; \
+        wr->dmember##_len = strlen(str); \
+        memcpy((uint8_t *)wr + offset, str, wr->dmember##_len + 1); \
+        offset += wr->dmember##_len + 1; \
+    } while (0)
+
+    r = req->request;
+    offset = sizeof(nxt_wasm_request_t)
+             + (r->fields_count * sizeof(nxt_wasm_http_hdr_field_t));
```

Comment:
For these things, offsetof(3) is more appropriate than sizeof().  It gets the offset of the flexible array member correctly in both `[0]` and `[1]` cases.

Reply:
I'm not sure about that in this case.

This isn't finding a member in a struct, it's setting the offset to after the end of the structure and that will vary depending on how many headers there were.

What we have is

nxt_wasm_request_t
+ 0 or more nxt_wasm_http_hdr_field_t
+ 0 or more header name/value pairs (described by the nxt_wasm_http_hdr_field_t structs)
+ 0 or more bytes of body content

In this case offset is pointing to after the nxt_wasm_http_hdr_field_t's

Reply:
We agreed to use sizeof().